src/persist-client/src/lib.rs

```diff
@@ -446,6 +447,178 @@ mod tests {
         );
     }
 
+    #[tokio::test]
+    async fn invalid_usage() {
+        mz_ore::test::init_logging();
+
+        let data = vec![
+            (("1".to_owned(), "one".to_owned()), 1, 1),
+            (("2".to_owned(), "two".to_owned()), 2, 1),
+            (("3".to_owned(), "three".to_owned()), 3, 1),
+        ];
+
+        let shard_id = ShardId::new();
+        let client = new_test_client().await;
+
+        let (mut write0, read0) = client
+            .expect_open::<String, String, u64, i64>(shard_id)
+            .await;
+
+        write0.expect_compare_and_append(&data, 0, 4).await;
+
+        // InvalidUsage from PersistClient methods.
+        {
+            assert_eq!(
+                format!(
+                    "{:?}",
+                    client.open::<Vec<u8>, String, u64, i64>(shard_id).await
+                ),
+                "Err(CodecMismatch { requested: (\"Vec<u8>\", \"String\", \"u64\", \"i64\"), actual: (\"String\", \"String\", \"u64\", \"i64\") })"
```

Comment:
Why are you using string comparison instead of something like:

```rust
            assert_eq!(
                client
                    .open::<Vec<u8>, String, u64, i64>(shard_id)
                    .await
                    .unwrap_err(),
                InvalidUsage::CodecMismatch {
                    requested: tpe("Vec<u8>", "String", "u64", "i64",),
                    actual: tpe("String", "String", "u64", "i64",),
                }
            );
```

Where `tpe()` is a helper I made up. Plus I had to add `#[cfg_attr(test, derive(PartialEq, Eq))]` on `InvalidUsage`.

The strings seem somewhat hard to maintain, but there probably is a good reason. :sweat_smile:

Reply:
I usually lean toward matching on error message in tests because that's often how they're consumed in production. unwrap_err is a good idea, I simply forgot it existed :). I'll switch to that!